service/history/api/replication/syncworkflowstate.go

```diff
@@ -0,0 +1,298 @@
+}
+
+func getEventsBlob(ctx context.Context, shardContext shard.Context, branchToken []byte, startEventId int64, endEventId int64) ([]*common.DataBlob, error) {
+	rawHistoryResponse, err := shardContext.GetExecutionManager().ReadRawHistoryBranch(ctx, &persistence.ReadHistoryBranchRequest{
+		BranchToken: branchToken,
+		MinEventID:  startEventId,
+		MaxEventID:  endEventId,
+		PageSize:    100,
+		ShardID:     shardContext.GetShardID(),
+	})
+	if err != nil {
+		return nil, err
+	}
+	return rawHistoryResponse.HistoryEventBlobs, nil
+}
+
+func getSyncStateEvents(ctx context.Context, shardContext shard.Context, targetVersionHistories *history.VersionHistories, sourceVersionHistories *history.VersionHistories, isSameBranch bool) ([]*common.DataBlob, error) {
+	startEventId := common2.EndEventID
+	sourceHistory, err := versionhistory.GetCurrentVersionHistory(sourceVersionHistories)
+	if err != nil {
+		return nil, err
+	}
+	if isSameBranch {
+		targetHistory, err := versionhistory.GetCurrentVersionHistory(targetVersionHistories)
+		result, err := versionhistory.CompareVersionHistory(targetHistory, sourceHistory)
+		if err != nil {
+			return nil, err
+		}
+		if result > 0 {
+			return nil, serviceerror.NewInvalidArgument("target version history greater than source version history")
+		}
+		if result == 0 {
+			return nil, nil
+		}
+		targetLastItem, err := versionhistory.GetLastVersionHistoryItem(targetHistory)
+		if err != nil {
+			return nil, err
+		}
+		startEventId = targetLastItem.GetEventId() + 1
+	} else {
+		lcaItem, _, err := versionhistory.FindLCAVersionHistoryItemAndIndex(targetVersionHistories, sourceHistory)
+		if err != nil {
+			return nil, err
+		}
+		startEventId = lcaItem.GetEventId() + 1
+	}
+
+	sourceLastItem, err := versionhistory.GetLastVersionHistoryItem(sourceHistory)
+	if err != nil {
+		return nil, err
+	}
+	return getEventsBlob(ctx, shardContext, sourceHistory.BranchToken, startEventId, sourceLastItem.GetEventId()+1)
+}
+
+func isInfoUpdated(subStateMachine LastUpdatedStateTransitionGetter, versionedTransition *persistencepb.VersionedTransition) bool {
+	if subStateMachine == nil {
+		return false
+	}
+	lastUpdate := subStateMachine.GetLastUpdateVersionedTransition()
+	return workflow.CompareVersionedTransition(lastUpdate, versionedTransition) > 0
+}
+
+func getUpdatedInfo[K comparable, V LastUpdatedStateTransitionGetter](subStateMachine map[K]V, versionedTransition *persistencepb.VersionedTransition) map[K]V {
+	result := make(map[K]V)
+	for k, v := range subStateMachine {
+		if isInfoUpdated(v, versionedTransition) {
+			result[k] = v
+		}
+	}
+	return result
+}
+
+func getUpdatedSubStateMachine(n *hsm.Node, versionedTransition *persistencepb.VersionedTransition) ([]*persistencepb.WorkflowMutableStateMutation_StateMachineNodeMutation, error) {
+	var updatedStateMachines []*persistencepb.WorkflowMutableStateMutation_StateMachineNodeMutation
+	walkFn := func(node *hsm.Node) error {
+		if node == nil {
```

Comment:
it should never be nil I think. If so, we should return an error.

We also need to exclude the root (node.parent == nil) here, which is mutable state itself.

Reply:
Done.